pkg/kubelet/images/image_manager.go

```diff
@@ -63,30 +94,73 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 	} else {
 		puller = newParallelImagePuller(imageService, maxParallelImagePulls)
 	}
+
+	stateImpl, err := NewCheckpointState(stateFileDirectory, secretPulledImageFileName)
+	// TODO (pacoxu) make a Noop state implementation for test
+	if err != nil {
+		klog.ErrorS(err, "Could not initialize checkpoint manager.")
+	}
+
 	return &imageManager{
 		recorder:               recorder,
 		imageService:           imageService,
 		backOff:                imageBackOff,
 		puller:                 puller,
 		podPullingTimeRecorder: podPullingTimeRecorder,
+		keyring:                keyring,
+		state:                  stateImpl,
 	}
 }
 
 // shouldPullImage returns whether we should pull an image according to
 // the presence and pull policy of the image.
-func shouldPullImage(container *v1.Container, imagePresent bool) bool {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret bool) bool {
 	if container.ImagePullPolicy == v1.PullNever {
 		return false
 	}
 
-	if container.ImagePullPolicy == v1.PullAlways ||
-		(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
-		return true
+	if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) { // ungated old behavior
+		if container.ImagePullPolicy == v1.PullAlways ||
+			(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
+			return true
+		}
+	} else { // new gated behavior
+		if container.ImagePullPolicy == v1.PullAlways {
+			return true
+		}
+		if container.ImagePullPolicy == v1.PullIfNotPresent {
+			if !imagePresent {
+				return true
+			}
+			// if the imageRef has been pulled by a secret and Pull Policy is PullIfNotPresent
+			// we need to ensure that the current pod's secrets map to an auth that has Already
+			// pulled the image successfully. Otherwise pod B could use pod A's images
+			// without auth. So in this case if pulledBySecret but not ensured by matching
+			// secret auth for a pull again for the pod B scenario where the auth does not match
+			if pulledBySecret && !ensuredBySecret {
+				return true // noting here that old behaviour returns false in this case indicating the image should not be pulled
+			}
+		}
 	}
-
 	return false
 }
 
+func (e *ensuredSecretPulledImageDigest) Clone() *ensuredSecretPulledImageDigest {
+	c := &ensuredSecretPulledImageDigest{
+		HashMatch: make(map[string]*EnsuredBySecret),
+	}
+	for k, v := range e.HashMatch {
+		c.HashMatch[k] = v
+	}
+	c.ImageName = e.ImageName
+	return c
+}
+
+func (m *imageManager) StartGC() {
+	klog.InfoS("Starting Kubelet image pull secret ensuring GC manager")
+	go wait.Until(func() { m.state.ClearExpiredState() }, 1*time.Minute, wait.NeverStop)
```

Comment:
seems aggressive to run this once a min... even if empty, even if the next timeout is closer to ensuredExpirationDuration = 1 * time.Hour 

Maybe if clear expired state returned the duration from now to the next timeout...

Reply:
+1 on making the GC interval 1 hour.